### DISCUSSION TITLE
Can't have generics on app state marked with FromRef?

### DISCUSSION THREAD

**Discussion by ghost** [tokio-rs/axum#1872] [Q&A] [answered]
I was trying to get dependency injection to work more or less nicely with generics, but I got kinda stomped before I could even begin. Let's say I have this

```rust
#[derive(Clone, FromRef)]
struct AppState<T: ApiClient> {
    thing: T,
}

trait ApiClient {
    fn do_the_thing(&self);
}
```

Rust then complains with the following error:

```
error[E0107]: missing generics for struct `AppState`
  --> src/main.rs:15:8
   |
15 | struct AppState<T: ApiClient> {
   |        ^^^^^^^^ expected 1 generic argument
   |
note: struct defined here, with 1 generic parameter: `T`
  --> src/main.rs:15:8
   |
15 | struct AppState<T: ApiClient> {
   |        ^^^^^^^^ -
help: add missing generic argument
   |
15 | struct AppState<T><T: ApiClient> {
   |                +++
```

Obviously `struct AppState<T><T: ApiClient>` is not valid syntax either. What the heck is going on?

**Comment by davidpdrsn** [comment, marked as answer]
Try writing a manual implementation.

**Reply by ghost** [reply]
Ohhhhh I see how it works now. I still get a different error if I try to implement it for `T: ApiClient`, but it works for a concrete implementation.

Thank you!

**Reply by davidpdrsn** [reply]
Yeah I think it conflicts with https://docs.rs/axum/latest/axum/extract/trait.FromRef.html#impl-FromRef%3CT%3E-for-T

**Reply by jplatte** [reply]
Maybe we should raise a more useful error when the derive is used on a generic type.

**Reply by davidpdrsn** [reply]
Yeah that would be good! I'll do a quick PR for that.
